### DISCUSSION TITLE
Switch the DateFormat from Date Time type

### DISCUSSION THREAD

**Discussion by LeonardoIPM** [eclipse-birt/birt#1852] [Q&A] [answered]
Is there some configuration to set the output date time format I want?
When updating BIRT version and the ICU4J lib from v53 to v75, our date formatting changed.

The reports that had date format like "dd/MM/yyyy" has become "dd 'de' MMM. 'de' yyyy, HH:mm".

This happens in Data Item, when selecting DataType = Date Time. 
Expression example: "16/01/2024 10:11:14.65088".

**Comment by speckyspooky** [comment]
Yes, you can do it on BIRT-level with the DateTime Format of the DataItem.
With custom you can use predefined formats or self defined formats.

![grafik](https://github.com/user-attachments/assets/8c7bd0c6-85b2-42f4-ae53-c58d383edf5d)

This option can be used on script level with the BIRT.formatter() as well.

(Or you can use at worst-case scenario SQL-functions.)

**Reply by LeonardoIPM** [reply]
Perfect.

Just one mor question, we have tons of reports, most of them was created in BIRT 4.4. If I open one of them in Birt client 4.4, then the format is setted as dd/MM/yyyy, but if I open with Birt client 4.15, then the format is setted as dd 'de' MMM. 'de' yyyy. 

Some default configuration was changed between versions? Maybe I can revert it working on reportDesign at code-level?

(In XML level, they are using the same Medium Date pattern, but seems that in newer versions the ICU4J changed the standard, the dd/MM/yyyy become Short Date).

**Reply by speckyspooky** [reply, marked as answer]
This is a very good question because my reports works in every case with specified formats.
So, yes I understand the topic and it would be hard to change all values. But currently I figured out no global setting to be able for a new global default of the date format :o(

**Reply by LeonardoIPM** [reply]
Thank you for the time, probably we will maintain the old versions of ICU until find a workaround.

**Reply by LeonardoIPM** [reply]
> This is a very good question because my reports works in every case with specified formats. So, yes I understand the topic and it would be hard to change all values. But currently I figured out no global setting to be able for a new global default of the date format :o(

So, we decided to recursively take the report elements and change the dateTimeFormat property from "Medium Date" to "Short Date". It worked, but unfortunately there are some reports with DateTime Format as "Unformatted" and the dateTimeFormat property isn't setted. 
Is there a way to take the Data Type from a DataItemHandle? This way I could set the property in Date or DateTime format even if it wasn't setted. Otherwise I didn't find a solution

**Reply by speckyspooky** [reply]
Yes, you can get the used format on the DataItem->onCreate(): `this.getStyle().dateFormat;`
For "unformatted" values the result is null but for the rest the date format style is placed there.

Attached the screens of my tests:

Test-parameter for output
![grafik](https://github.com/user-attachments/assets/65f5f7b0-074c-415c-9f19-4d9bf33bd909)

DateFormat unset:
![grafik](https://github.com/user-attachments/assets/d0576c1c-9bb2-4393-a3d8-ce943b25f57a)

DateFormat set:
![grafik](https://github.com/user-attachments/assets/d10d8945-83fd-4174-9f96-17972b206c8e)

**Comment by LeonardoIPM** [comment]
Actually, what i want is to get the dataType property using the DesignEngine API... But it doesn't seems to be possible. I can recover the dateFormat property of an element, but not the dataType... Then, if a DateFormat is undefined, I just can't discover him and it becomes a Medium Date format.

Idk if would be viable to create a configuration to change the default value in /org.eclipse.birt.core/src/org/eclipse/birt/core/format/DateFormatter.java class
```
if (formatString == null || UNFORMATTED.equals(formatString)) {
	formatPattern = UNFORMATTED;
	dateTimeFormat = com.ibm.icu.text.DateFormat.getDateTimeInstance(com.ibm.icu.text.DateFormat.MEDIUM,
			com.ibm.icu.text.DateFormat.SHORT, locale);
	dateTimeFormat = toPredefinedPattern((SimpleDateFormat) dateTimeFormat, locale);
	dateFormat = com.ibm.icu.text.DateFormat.getDateInstance(com.ibm.icu.text.DateFormat.MEDIUM, locale);
	timeFormat = com.ibm.icu.text.DateFormat.getTimeInstance(com.ibm.icu.text.DateFormat.MEDIUM, locale);
	return;
```

If instead a DateFormat.MEDIUM it was configurable, it would resolve my issue.

**Reply by speckyspooky** [reply]
Ok, I understand. Another suggestion, you can define for each report a standard CSS-style with the name "report".
And there you can change the default of the DateTime like a general setting of the report.

CSS-sytle "report":
![grafik](https://github.com/user-attachments/assets/ac38a561-672b-4177-8a7a-7a8538b0cdbc)

CSS datetime format:
![grafik](https://github.com/user-attachments/assets/a7acd72f-9514-4f0e-a632-6babd020ce8a)

**Reply by LeonardoIPM** [reply]
Perfect, I think this will resolve my issue. Thank you @speckyspooky !

**Reply by speckyspooky** [reply]
You are welcome.
